src/modules/notifications/services/notifications.service.ts

```diff
@@ -0,0 +1,101 @@
+import { Injectable, Logger, NotFoundException } from '@nestjs/common';
+import { IPagination, Pagination } from '@studiohyperdrive/pagination';
+import { get } from 'lodash';
+
+import { GqlCreateOrUpdateNotification, GqlNotification, Notification } from '../types';
+
+import {
+	FIND_NOTIFICATIONS_BY_USER,
+	INSERT_NOTIFICATION,
+	UPDATE_NOTIFICATION,
+} from './queries.gql';
+
+import { DataService } from '~modules/data/services/data.service';
+import { PaginationHelper } from '~shared/helpers/pagination';
+
+@Injectable()
+export class NotificationsService {
+	private logger: Logger = new Logger(NotificationsService.name, { timestamp: true });
+
+	constructor(protected dataService: DataService) {}
+
+	/**
+	 * Adapt a notification as returned by a typical graphQl response to our internal notification data model
+	 */
+	public adaptNotification(
+		gqlNotification: GqlNotification | undefined
+	): Notification | undefined {
+		if (!gqlNotification) {
+			return undefined;
+		}
+		return {
+			id: get(gqlNotification, 'id'),
+			description: get(gqlNotification, 'description'),
+			title: get(gqlNotification, 'title'),
+			status: get(gqlNotification, 'status'),
+			visitId: get(gqlNotification, 'visit_id'),
+			createdAt: get(gqlNotification, 'created_at'),
+			updatedAt: get(gqlNotification, 'updated_at'),
+			notificationType: get(gqlNotification, 'notification_type'),
+			showAt: get(gqlNotification, 'show_at'),
+		};
+	}
+
+	public async findNotificationsByUser(
+		userProfileId: string,
+		moreRecentThan: string,
+		page = 1,
+		size = 20
+	): Promise<IPagination<Notification>> {
+		const { offset, limit } = PaginationHelper.convertPagination(page, size);
+		const notificationsResponse = await this.dataService.execute(FIND_NOTIFICATIONS_BY_USER, {
+			userProfileId,
+			moreRecentThan,
+			offset,
+			limit,
+		});
+
+		return Pagination<Notification>({
+			items: notificationsResponse.data.app_notification.map((notification: any) =>
+				this.adaptNotification(notification)
+			),
+			page,
+			size,
+			total: notificationsResponse.data.app_notification_aggregate.aggregate.count,
+		});
+	}
+
+	public async create(
+		notification: Partial<GqlCreateOrUpdateNotification>
+	): Promise<Notification> {
+		const response = await this.dataService.execute(INSERT_NOTIFICATION, {
+			object: notification,
+		});
+		const createdNotification = response?.data?.insert_app_notification?.returning[0];
+		this.logger.debug(`Notification ${createdNotification?.id} created`);
+
+		return this.adaptNotification(createdNotification);
+	}
+
+	public async update(
+		notificationId: string,
+		userProfileId: string,
+		notification: Partial<GqlCreateOrUpdateNotification>
+	): Promise<Notification> {
+		const response = await this.dataService.execute(UPDATE_NOTIFICATION, {
+			notificationId,
+			userProfileId,
+			notification,
+		});
+
+		const updatedNotification = response?.data?.update_app_notification?.returning?.[0];
```

Comment:
response?.data?.update_app_notification? gaat altijd bestaan; maar update_app_notification can null zijn
Kan je schrijven als:
`const { data: { update_app_notification: updatedNotification } } = await this.dataService ... `

Reply:
maar dan zijn jouw updatedNotification en mijn updatedNotification toch niet hetzelfde?
aangezien ik dan nog .returning?.[0] uit de updatedNotification moet halen.

Het lijkt me cleaner om het volledige pad op 1 plaats te zetten, dan het deels met een deconstructor te doen en een deel met dot notatie.

Reply:
nee, maar als `updatedNotification` daar null is dan bestaat het niet en throw je exception, en anders bestaat het zeker...
de gustibus et coloribus... maar ik vind dat cleaner :)